### PR TITLE
Symex Documentation

### PR DESCRIPTION
**Pull request by NathanJPhillips**
More could be done on this but this is an improvement on the current lack of documentation.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/symex_main.cpp [hunk 1]
```diff
@@ -229,6 +229,12 @@ void goto_symext::symex_threaded_step(
   }
 }
 
+/// Create a default delegate to retrieve function bodies from a
+/// goto_functionst
+/// \param goto_functions The function map from which to retrieve function
```
Comment (tautschnig) [inline src/goto-symex/symex_main.cpp]: colon after parameter name

File: src/goto-symex/goto_symex.h [hunk 2]
```diff
@@ -38,38 +38,59 @@ class namespacet;
...
 /// documented in the \ref symex-overview
 /// "Symbolic execution module page".
 class goto_symext
 {
 public:
+  /// A type abbreviation for \ref goto_symex_statet
   typedef goto_symex_statet statet;
 
+  /// Construct a goto_symext to execute a particular program
+  /// \param mh The message handler to use for log messages
```
Comment (tautschnig) [inline src/goto-symex/goto_symex.h]: colon after parameter (here and everywhere else)

File: src/goto-symex/goto_symex.h [hunk 3]
```diff
@@ -91,18 +111,28 @@ class goto_symext
...
-  /// around afterwards.
+  /// \remarks
+  /// The state that goto_symext maintains uses a lot of memory.
+  /// This method therefore deallocates the state as soon as symbolic execution
+  /// has completed. This function is useful to callers that don't care about
+  /// having the state around afterwards.
+  /// \param get_goto_function The delegate to retrieve function bodies (see
+  /// \ref get_goto_functiont)
+  /// \param new_symbol_table A symbol table to store the symbols added by
+  /// SymEx
```
Comment (tautschnig) [inline src/goto-symex/goto_symex.h]: This camel-casing makes it sound like there were one particular thing of that name. Maybe just say "store the symbols adding during symbolic execution" (applies here and elsewhere)

File: src/goto-symex/goto_symex.h [hunk 4]
```diff
@@ -151,16 +195,18 @@ class goto_symext
   bool should_pause_symex;
 
 protected:
+  /// The configuration to use for this symbolic execution
   const symex_configt symex_config;
 
-  /// Initialise the symbolic execution and the given state with <code>pc</code>
+  /// Initialise the given symbolic execution state with <code>pc</code>
```
Comment (tautschnig) [inline src/goto-symex/goto_symex.h]: Using `\p pc` is shorter

File: src/goto-symex/goto_symex.h [hunk 5]
```diff
@@ -151,16 +195,18 @@ class goto_symext
...
-  /// \param function_identifier: The function in which the instructions are
-  /// \param pc: first instruction to symex
-  /// \param limit: final instruction, which itself will not
-  ///   be symexed.
+  /// \param state: SymEx state to initialise.
+  /// \param get_goto_function: The delegate to retrieve function bodies (see
+  /// \ref get_goto_functiont)
+  /// \param function_identifier: The identifier of the function containing the
+  /// instructions
+  /// \param pc: The first instruction to execute
```
Comment (tautschnig) [inline src/goto-symex/goto_symex.h]: Maybe add "(program counter)" as that was a recent source of discussion.

File: src/goto-symex/goto_symex.h [hunk 6]
```diff
@@ -385,22 +517,62 @@ class goto_symext
...
-  virtual void symex_input(statet &, const codet &);
-  virtual void symex_output(statet &, const codet &);
-
+    statet &state, const exprt &lhs, const side_effect_exprt &code);
+  /// Symbolically execute a FUNCTION_CALL instruction for a function whose
+  /// name starts with CPROVER_FKT_PREFIX
+  /// \remarks
+  /// While the name seems to imply that this would be called when SymEx
+  /// doesn't know what to do, it may actually be derived from a German
+  /// abbreviation for function.
```
Comment (tautschnig) [inline src/goto-symex/goto_symex.h]: Looool. It had never occurred to me that this was not an appropriate English abbreviation at all... (But also I think it's just something in TG blocking us from removing it completely.)

**Review by tautschnig — APPROVED** [review]

**Comment by owen-mc-diffblue** [comment]
The `\brief`s are redundant. The first sentence is taken as the brief description when the doxygen is compiled. You're free to keep them if you prefer to have them there.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c2f5c9b4e5aea5bd25d4e8bd21aa719894c818cd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102907250

**Pull request merged**